Redo tempo-cli with basic command structure and improvements (#385)
grafana/tempo: docs/tempo/website/cli/_index.md

---
title: "Tempo CLI"
description: "Guide to using tempo-cli"
keywords: ["tempo", "cli", "tempo-cli", "command line interface"]
weight: 450
---

# Tempo CLI

> Tempo CLI is a separate executable that contains utility functions related to tempo.  It is not required for a working install, but may be helpful for analysis or troubleshooting.

Reviewer: ```suggestion
Tempo CLI is a separate executable that contains utility functions related to the tempo software.  Although it is not required for a working installation, Tempo CLI can be helpful for deeper analysis or for troubleshooting.
```

## Tempo CLI command syntax

The general syntax for commands in Tempo CLI is:
```bash
tempo-cli command [subcommand] [options] [arguments...]
```

> 

Reviewer: ```suggestion
Remove extra space.
```

`--help` or `-h` displays the help for a command or subcommand.

**Example:**
```bash
tempo-cli -h
tempo-cli command [subcommand] -h
```

## Running Tempo CLI

> Tempo CLI is currently only available as source and a working go installation is required to build it.  It can be compiled to a native binary and executed normally, or it can be executed using the `go run` command. 

Reviewer: ```suggestion
Tempo CLI is currently available as a source. A working Go installation is required to build it. It can be compiled to a native binary and executed normally, or it can be executed using the `go run` command. 
```

Author: Updated as discussed

**Example:**
```bash
./tempo-cli [arguments...]
go run ./cmd/tempo-cli [arguments...]
```

## Backend options

> Tempo CLI connects directly to the storage backend for some commands, meaning it requires the ability to read from S3, GCS, or file-system storage.  The backend can be configured in a few ways:

Reviewer: ```suggestion
Tempo CLI connects directly to the storage backend for some commands, meaning that it requires the ability to read from S3, GCS, or file-system storage.  The backend can be configured in a few ways:
```

> * Load an existing tempo config file using `--config-file` (`-c`) option. This is the recommended option for frequent usage. See [Configuration](../configuration/) documentation for more information.

Reviewer: ```suggestion
* Load an existing tempo configuration file using the `--config-file` (`-c`) option. This is the recommended option for frequent usage. Refer to [Configuration](../configuration/) documentation for more information.
```

> * Specify the individual settings:

Reviewer: ```suggestion
* Specify individual settings:
```

>     * `--backend <value>` The storage backend type, one of `s3`, `gcs`, `local`

Reviewer: ```suggestion
    * `--backend <value>` The storage backend type, one of `s3`, `gcs`, and `local`.
```

>     * `--bucket <value>` The bucket name.  The meaning of this value is backend specific. See configuration documentation for more information.

Reviewer: ```suggestion
    * `--bucket <value>` The bucket name. The meaning of this value is backend-specific. Refer to [Configuration](../configuration/) documentation for more information.

    * `--s3-endpoint <value>` The S3 API endpoint (i.e. s3.dualstack.us-east-2.amazonaws.com).

>     * `--s3-user <value>`, `--s3-password <value>` The S3 user name and password (or access key and secret key). Optional, tempo-cli supports the same authentication mechanisms as tempo. See [S3 permissions documentation](../configuration/s3/#permissions) for more information.

Reviewer: ```suggestion
    * `--s3-user <value>`, `--s3-password <value>` The S3 user name and password (or access key and secret key). Optional, as Tempo CLI supports the same authentication mechanisms as Tempo. See [S3 permissions documentation](../configuration/s3/#permissions) for more information.
```

Each option applies only to the command in which it is used. For example, `--backend <value>` does not permanently change where Tempo stores data. It only changes it for command in which you apply the option.

## Query Command
Call the tempo API and retrieve a trace by ID.
```bash
tempo-cli query <api-endpoint> <trace-id>
```

Arguments:

> - `api-endpoint` Url for tempo api

Reviewer: ```suggestion
- `api-endpoint` URL for tempo API.
```

> - `trace-id` Trace ID as hexadecimal string

Reviewer: ```suggestion
- `trace-id` Trace ID as a hexadecimal string.

Options:

> - `--org-id <value>` Organization ID (for use in multi-tenant setup)

Reviewer: ```suggestion
- `--org-id <value>` Organization ID (for use in multi-tenant setup).
```

**Example:**
```bash
tempo-cli query http://tempo:3100 f1cfe82a8eef933b
```

## List Blocks
Lists information about all blocks for the given tenant, and optionally perform integrity checks on indexes for duplicate records.

```bash
tempo-cli list blocks <tenant-id>
```

Arguments:
- `tenant-id` The tenant ID.  Use `single-tenant` for single tenant setups.

Options:

> - `--load-index` Also load the block indexes and perform integrity checks for duplicates. _Warning:_ can be intense.

Reviewer: ```suggestion
- `--load-index` Also load the block indexes and perform integrity checks for duplicates. 
_Warning:_ can be intense.
```

**Output:**
Explanation of output:

> - `ID` block ID
> - `Lvl` Compaction level of the block
> - `Count` Number of objects stored in the block
> - `Window` The time window as considered for compaction purposes
> - `Start` The earliest timestamp stored in the block
> - `End` The latest timestamp stored in the block
> - `Duration` Time duration between start and end
> - `Idx` Number of records stored in the index (present when --load-index is specified)
> - `Dupe` Number of duplicate entries in the index. Should be zero.  (present when --load-index is specified)

Reviewer: ```suggestion
- `ID` Block ID.
- `Lvl` Compaction level of the block.
- `Count` Number of objects stored in the block.
- `Window` The time window considered for compaction purposes.
- `Start` The earliest timestamp stored in the block.
- `End` The latest timestamp stored in the block.
- `Duration` Time duration between start and end.
- `Idx` Number of records stored in the index (present when --load-index is specified).
- `Dupe` Number of duplicate entries in the index. Should be zero.  (present when --load-index is specified).
```

**Example:**
```bash
tempo-cli list blocks -c ./tempo.yaml single-tenant
```

## List Block
Lists information about a single block, and optionally, perform an integrity check for duplicate records.

```bash
tempo-cli list block <tenant-id> <block-id>
```

Arguments:
- `tenant-id` The tenant ID.  Use `single-tenant` for single tenant setups.
- `block-id` The block ID as UUID string.

Options:

> - `--check-dupes` Also load the block data and perform integrity check for duplicates. _Warning:_ can be intense.

Reviewer: ```suggestion
- `--check-dupes` Also load the block data and perform an integrity check for duplicates. 
_Warning:_ can be intense.
```

Author: As discussed, this was changed to **Note**
